pkg/controller/data_upload_controller.go

```diff
@@ -0,0 +1,498 @@
+	du := &velerov2alpha1api.DataUpload{}
+	err := r.client.Get(context.Background(), types.NamespacedName{
+		Namespace: pod.Namespace,
+		Name:      pod.Labels[velerov1api.DataUploadLabel],
+	}, du)
+
+	if err != nil {
+		r.logger.WithField("Backup pod", pod.Name).WithError(err).Error("unable to get dataupload")
+		return []reconcile.Request{}
+	}
+
+	if du.Status.Phase != velerov2alpha1api.DataUploadPhaseAccepted {
+		return []reconcile.Request{}
+	}
+
+	r.logger.WithField("Backup pod", pod.Name).Infof("Preparing dataupload %s", du.Name)
+	if err := r.patchDataUpload(context.Background(), du, prepareDataUpload); err != nil {
+		r.logger.WithField("Backup pod", pod.Name).WithError(err).Error("failed to patch dataupload")
+		return []reconcile.Request{}
+	}
+
+	requests := reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Namespace: du.Namespace,
+			Name:      du.Name,
+		},
+	}
+	return []reconcile.Request{requests}
+}
+
+func (r *DataUploadReconciler) patchDataUpload(ctx context.Context, req *velerov2alpha1api.DataUpload, mutate func(*velerov2alpha1api.DataUpload)) error {
+	original := req.DeepCopy()
+	mutate(req)
+	if err := r.client.Patch(ctx, req, client.MergeFrom(original)); err != nil {
+		return errors.Wrap(err, "error patching DataUpload")
+	}
+
+	return nil
+}
+
+func prepareDataUpload(du *velerov2alpha1api.DataUpload) {
+	du.Status.Phase = velerov2alpha1api.DataUploadPhasePrepared
+}
+
+func (r *DataUploadReconciler) errorOut(ctx context.Context, du *velerov2alpha1api.DataUpload, err error, msg string, log logrus.FieldLogger) (ctrl.Result, error) {
+	if se, ok := r.snapshotExposerList[du.Spec.SnapshotType]; ok {
+		se.CleanUp(ctx, getOwnerObject(du), du.Spec.CSISnapshot.VolumeSnapshot, du.Spec.SourceNamespace)
+	} else {
+		err = errors.Wrapf(err, "failed to clean up exposed snapshot with could not find %s snapshot exposer", du.Spec.SnapshotType)
+	}
+	return ctrl.Result{}, r.updateStatusToFailed(ctx, du, err, msg, log)
+}
+
+func (r *DataUploadReconciler) updateStatusToFailed(ctx context.Context, du *velerov2alpha1api.DataUpload, err error, msg string, log logrus.FieldLogger) error {
+	original := du.DeepCopy()
+	du.Status.Phase = velerov2alpha1api.DataUploadPhaseFailed
+	du.Status.Message = errors.WithMessage(err, msg).Error()
+	if du.Status.StartTimestamp.IsZero() {
+		du.Status.StartTimestamp = &metav1.Time{Time: r.clock.Now()}
+	}
+
+	du.Status.CompletionTimestamp = &metav1.Time{Time: r.clock.Now()}
+	if err = r.client.Patch(ctx, du, client.MergeFrom(original)); err != nil {
+		log.WithError(err).Error("error updating DataUpload status")
+		return err
+	}
+
+	return nil
+}
+
+func (r *DataUploadReconciler) acceptDataUpload(ctx context.Context, du *velerov2alpha1api.DataUpload) (bool, error) {
+	updated := du.DeepCopy()
+	updated.Status.Phase = velerov2alpha1api.DataUploadPhaseAccepted
+
+	r.logger.Infof("Accepting snapshot backup %s", du.Name)
+
+	time.Sleep(2 * time.Second)
+
+	err := r.client.Update(ctx, updated)
```

Comment:
It seems this line uses `Update` method on-purpose to make sure the request failed on conflict.
If that's correct, please add some comments to explain the purpose.